aten/src/ATen/native/Unique.cpp

```diff
@@ -47,6 +47,81 @@ std::tuple<Tensor, Tensor> _unique_cpu_template(
   }
   return std::make_tuple(output, inverse_indices);
 }
+
+template <typename scalar_t>
+std::tuple<Tensor, Tensor> _unique_dim_cpu_template(
+    const Tensor& self,
+    const int64_t dim,
+    const bool return_inverse) {
+  // reshape tensor as [dim, -1]
+  Tensor input_flat = self.transpose(dim, 0);
+  std::vector<int64_t> orig_sizes(input_flat.sizes().begin(), input_flat.sizes().end());
+  input_flat = input_flat.contiguous().view({input_flat.size(0), -1});
+
+  std::vector<int64_t> indices(input_flat.size(0));
+  std::iota(indices.begin(), indices.end(), 0);
+  int64_t numel = input_flat.size(1);
+  scalar_t* input_flat_ptr = ((scalar_t*)input_flat.data_ptr());
+
+  // sort indices using data
+  std::sort(indices.begin(), indices.end(),
+    [&](int64_t a, int64_t b) -> bool {
+      for (int64_t i = 0; i < numel; ++i) {
+        scalar_t lhs = input_flat_ptr[i + a * numel];
+        scalar_t rhs = input_flat_ptr[i + b * numel];
+        if (lhs < rhs) {
+          return true;
+        } else if (lhs > rhs) {
+          return false;
+        }
+      }
+      return false;
+    });
+
+  Tensor input_sorted = at::empty(input_flat.sizes(), input_flat.type());
+  for (int i = 0; i < indices.size(); ++i) {
+    input_sorted[i] = input_flat[indices[i]];
+  }
+ 
+  // pre-calculate mask for inverse_indices
+  Tensor mask = at::empty(input_sorted.size(0), self.type().toScalarType(kLong));
+  mask[0] = 1;
+  int mask_idx = 1;
+
+  std::vector<Tensor> input_unbind = at::unbind(input_sorted, 0);
+  auto last = std::unique(input_unbind.begin(), input_unbind.end(), [&](Tensor a, Tensor b) {
```

Comment:
I think that manually writing the equivalent loop for computing unique values and computing inverse_indices as part of that loop is going to be faster.

(As you pointed out earlier, that's not going to work for CUDA)

There's an example implementation of unique here:
https://en.cppreference.com/w/cpp/algorithm/unique

Reply:
Thanks for the suggestion. The `mask` used for the `inverse_indices` is already being pre-computed in the `std::unique` lambda function. 
I'll have a look at the implementation and will try to re-use the logic to manually write the loop.

Would you like to compute the `inverse_indices` as part of this loop without the mask as it's currently done?

Reply:
Yes, without the mask and the cumsum call.

There are two reasons for this:
1) I think it will be faster and may be less code
2) The current implementation assumes a certain implementation of std::unique that may not be true. (That unique traverses input_unbind once in order). While this is probably true for all implementations, I don't think it's guaranteed by the standard.

Reply:
The CUDA version is fine: the CUDA lambda doesn't modify anything. (It doesn't have this `mask_idx` counter)